vectorscope: add log scaling

dtorop opened the pull request:
As per suggestion from @TurboGit and @Mark-64. This makes the graph of low chroma images more legible.

There is a button to flip between linear & logarithmic graphing of vectorscope. Moved the colorspace button one to the right, to allow for linear/log button to be in the same place as in the regular histogram. Updated keyboard shortcut code to cycle through vectorscope variants (linear/log and uv/AzBz).

Updated hue ring graph to draw nodes at primary/secondary colors, as in log mode the angles at primaries/secondaries are less visible. Also tuned up graph scale circles.

Removed auto-scaling code, which was buggy and should be obviated by this.

Should fix #8813. Supersedes #8855, which I will close if this PR is a better route. May help with #8858, pending testing.

The log base for scaling is currently fixed via a `#define VECTORSCOPE_BASE_LOG 30`. It would be possible to add a UI to change this (e.g. via shift-scroll, if I could get `dt_gui_get_scroll_unit_delta()` to work on MacOS and Windows) which could obviate the linear/logarithmic button. But it would be nice if a simple button UI were enough.

## Examples

### Low chroma image

Adobe RGB histogram profile

#### linear scale

![image](https://user-images.githubusercontent.com/2311860/117392854-e7275300-aec0-11eb-9f63-df112b1918fd.png)

#### log scale

![image](https://user-images.githubusercontent.com/2311860/117392906-fc9c7d00-aec0-11eb-95f6-d4224ad72f5f.png)

### High chroma image

sRGB histogram profile

#### linear scale

![image](https://user-images.githubusercontent.com/2311860/117393260-bf84ba80-aec1-11eb-94dd-82690e324914.png)

### log scale

![image](https://user-images.githubusercontent.com/2311860/117393286-cb707c80-aec1-11eb-931f-76899245bac7.png)

MStraeten commented:
thanks, the log implementation fits to a couple of demands:
- similarity to common round vectorscorpe impelmentation in video editing (ok, thats just  user experience)
- mitigates the need to autoscale stuff
- mitigates the need for manual scaling since the height is properly used due to the near cyclic shape  - I'm not sure if an option for manual change of the log base is necessary (its also not there in log histogram) 
- mitigates the demand for recentering in a consequence

in my opinion the mitigations are near to "no longer necessary"

dtorop commented:
@MStraeten Great. I think it's the best so far as well.

> in my opinion the mitigations are near to "no longer necessary"

Likewise here.

TurboGit commented:
No matter what I do the histogram stays empty:

![image](https://user-images.githubusercontent.com/467069/117540197-3dee7300-b00e-11eb-8004-b9eb2423df76.png)

I have switched images, change histogram mode, selected log/linear and switched the color space, cleared my darktablerc... All scopes are displaying data except the vector scope for me.

dtorop commented:
> No matter what I do the histogram stays empty:
> 
> I have switched images, change histogram mode, selected log/linear and switched the color space, cleared my darktablerc... All scopes are displaying data except the vector scope for me.

@TurboGit: I think that this is because the new code can only handle matrix profiles (not CLUT), and is therefore failing if your display profile is a CLUT profile. A way to check should be to go to "display profile" via the softproof/gamut buttons on the bottom panel, and change to something which is definitely a matrix profile, e.g. "sRGB (web-safe)".

I'll see if there is a reasonable fix for this...

TurboGit commented:
Otherwise the log scale looks really nice and reads fine for most pictures I have tired. Good job !

dtorop commented:
> Worth trying to fix it as my screen is calibrated and using the profile from DisplayCAL makes the vectorscope unusable which is not good

I'll work on a fix, probably simply to roll back 955ee4b3e7136454ffc555539aa8cbfd134f1a20 rather than write code to use LCMS for CLUT profiles.

As an editorial for matrix profiles, I think they're about 4x faster than LUT, if DisplayCAL can be set to generate them well ("Curves + matrix" for profile type, probably).

This problem really goes back to the discussion in #8051 at least. The histogram runs after colorout on preview pipe, and hence sees data in display color profile rather than working (or export) profile. Hence there is an extra color profile hop:

`working profile -> display profile -> histogram profile`

What makes it worse is that the non-preview image doesn't make the hop through display profile. The display profile hop could either produce rounding errors or distort the color (perceptual intent).

Also, it's a waste of memory to allocate an approx. 19MB temporary buffer to hold the `display -> histogram` step then throw away that buffer right after the data is binned.

In the case of vectorscope, this is all extra-useless, as vectorscope works on chromaticity data in PCS, so these two (except for rounding errors) should be equivalent:

1. `working profile -> display profile -> histogram profile -> PCS`
2. `working profile -> display profile -> PCS`

The idea behind 955ee4b3e7136454ffc555539aa8cbfd134f1a20 was to make vectorscope lose the histogram profile hop (look more like 2. than 1.).

The right thing to do would probably be both:

1. Write something in `iop_profile.h` like `dt_ioppr_transform_image_colorspace_rgb_with_cb()` with a callback to the binning routine, but I don't think that would work without a bunch of inline functions. And there would need to be two cases, one transforming to PCS (for vectorscope) and one to RGB (for histogram/waveform).
2. Move histogram (and all other modules of interest to before colorout).

This is all way beyond the scope of this PR and probably v3.6, hence I think it's best to just revert that particular commit.

> Otherwise the log scale looks really nice and reads fine for most pictures I have tired. Good job !

Thank you! I feel happy about this version. I should have listened to the suggestion a while back!

TurboGit reviewed (approved):
Ok working fine for me now. This is ok for 3.6 will merge now. Thanks!
